src/parser/jobs/whm/modules/Lilies.js

```diff
@@ -0,0 +1,191 @@
+					<Trans id="whm.lily-cap.suggestion.content">
+						Use <ActionLink {...ACTIONS.AFFLATUS_SOLACE} /> or <ActionLink {...ACTIONS.AFFLATUS_RAPTURE} /> regularly to avoid capping your lilies.
+					</Trans>
+				</Fragment>,
+				severity: this._unusedLilies >= UNUSED_LILIES_MAX_MINOR ? SEVERITY.MEDIUM : SEVERITY.MINOR,
+				why: <Fragment>
+					<Trans id="whm.lily-cap.suggestion.why">
+						{<Plural value={this._unusedLilies} one="# lily" other="# lilies" />} lost to lily cap.
+					</Trans>
+				</Fragment>,
+			}))
+		}
+	}
+
+	checkBloodLilies() {
+		const possibleBloodLilies = (this._totalLilies / BLOOD_LILY_GENERATION)
+		console.log('possible = ', possibleBloodLilies, this._totalLilies)
+		this._missedBloodLilies = possibleBloodLilies - this._totalBloodLilies
+		const bloodWarnTarget = (possibleBloodLilies - 1) / possibleBloodLilies * 100
+		const bloodFailTarget = (possibleBloodLilies - WASTED_LILIES_MAX_MEDIUM) / possibleBloodLilies * 100
+
+		this.checklist.add(new TieredRule({
+			name: <Trans id="whm.lily-blood.checklist.name">
+				Spend your Blood Lily
+			</Trans>,
+			description: <Trans id="whm.lily-blood.checklist.content">
+				Never use <ActionLink {...ACTIONS.AFFLATUS_SOLACE} /> or <ActionLink {...ACTIONS.AFFLATUS_RAPTURE} /> when <ActionLink {...ACTIONS.AFFLATUS_MISERY} /> is available.  You lose DPS, and may miss the generation of additional Blood Lilies.
+			</Trans>,
+			tiers: {[bloodWarnTarget]: TARGET.WARN, [bloodFailTarget]: TARGET.FAIL, 100: TARGET.SUCCESS},
+			requirements: [
+				new Requirement({
+					name: <Trans id="whm.lily-blood.checklist.requirement.name">
+						Blood Lilies Generated
+					</Trans>,
+					value: this._totalBloodLilies,
+					target: Math.floor(possibleBloodLilies),
+				}),
+			],
+		}))
+
+		if (this._missedBloodLilies >= 1) {
+			this.suggestions.add(new Suggestion({
+				icon: ACTIONS.AFFLATUS_MISERY.icon,
+				content: <Fragment>
+					<Trans id="whm.lily-blood.suggestion.content">
+						Use <ActionLink {...ACTIONS.AFFLATUS_MISERY} /> to avoid wasting blood lily growth.
+					</Trans>
+				</Fragment>,
+				severity: this._missedBloodLilies >= WASTED_LILIES_MAX_MEDIUM ? SEVERITY.MAJOR : SEVERITY.MEDIUM,
+				why: <Fragment>
+					<Trans id="whm.lily-blood.suggestion.why">
+						<Plural value={this._missedBloodLilies} one="# blood lily" other="# blood lilies" /> did not bloom due to early lily use.
+					</Trans>
+				</Fragment>,
+			}))
+		}
+	}
+
+	/**
+	 * Calculates the number of lilies the user should currently have.
+	 *
+	 * Because lilies are generated every 30 seconds, and we don't want to
+	 * be constantly checking, we run a while loop every time the event is
+	 * more recent than when we should have gained a lily to catch up on our
+	 * lily generation.
+	 *
+	 * @param {*} event
+	 */
+	_calculateLilies(event) {
+		if (this._lastGained === 0) { this._lastGained = this.parser.fight.start_time }
+		while (event.timestamp > this._lastGained + LILY_INTERVAL) {
```

Comment:
Hm - while I doubt the loop will execute more than 0-1 times per event, I'm not particularly comfortable with it nonetheless.

Should be possible to calculate directly from the delta, something like

```js
// untested, batteries not included. manufacturer provides no warranty for suitability in a production environment
const delta = event.timestamp - this._lastGained
const newLilies = Math.floor(delta / LILY_INTERVAL)
this._lastGained += newLilies * LILY_INTERVAL

// modify _lilies/_totalLilies/_unusedLilies here etc
```

Reply:
fixed